File: app/src/main/java/com/ktc/sitepulse/domain/DateUtils.kt
package com.ktc.sitepulse.domain

import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.YearMonth
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * Mirrors the original app's date conventions exactly:
 *  - "today" for attendance doc IDs is the UTC calendar date
 *    (JS: new Date().toISOString().slice(0,10)) — NOT the device's local date.
 *  - Night-shift classification uses the LOCAL device hour.
 * Both are intentionally kept as in the web app so data stays consistent
 * with existing Firestore records.
 */
object DateUtils {
    private val ISO_DATE: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE

    fun todayStrUtc(): String = LocalDate.now(ZoneOffset.UTC).format(ISO_DATE)

    fun nowIso(): String = Instant.now().toString()

    fun yesterdayOf(dateStr: String): String =
        LocalDate.parse(dateStr, ISO_DATE).minusDays(1).format(ISO_DATE)

    fun localHour(): Int = LocalTime.now().hour

    /** nightStart/dayStart let a specific site override the company-wide 18:00/05:00 default
     * (e.g. a site running a genuinely different shift pattern) — see Site.nightStartHour. */
    fun shiftFor(hour: Int = localHour(), nightStart: Int = 18, dayStart: Int = 5): String =
        if (hour >= nightStart || hour < dayStart) "Night" else "Day"

    /** Formats an ISO instant string to a localized "h:mm a" time, or "" if null/unparseable. */
    fun formatTimeHm(iso: String?): String {
        if (iso.isNullOrBlank()) return ""
        return try {
            val instant = Instant.parse(iso)
            val local = LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
            local.format(DateTimeFormatter.ofPattern("h:mm a"))
        } catch (e: DateTimeParseException) {
            ""
        }
    }

    fun formatDateTime(iso: String?): String {
        if (iso.isNullOrBlank()) return ""
        return try {
            val instant = Instant.parse(iso)
            val local = LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
            local.format(DateTimeFormatter.ofPattern("d MMM yyyy, h:mm a"))
        } catch (e: DateTimeParseException) {
            ""
        }
    }

    fun monthOf(dateStr: String): String = dateStr.substring(0, 7) // "YYYY-MM"

    fun monthStart(monthStr: String): String = "$monthStr-01"

    fun monthEndExclusive(monthStr: String): String {
        val ym = YearMonth.parse(monthStr, DateTimeFormatter.ofPattern("yyyy-MM"))
        val next = ym.plusMonths(1)
        return "%04d-%02d-01".format(next.year, next.monthValue)
    }

    fun daysInMonth(monthStr: String): Int =
        YearMonth.parse(monthStr, DateTimeFormatter.ofPattern("yyyy-MM")).lengthOfMonth()

    /** Days elapsed so far if monthStr is the current UTC month, else the full month length. */
    fun elapsedDaysFor(monthStr: String): Int {
        val currentMonth = monthOf(todayStrUtc())
        return if (monthStr == currentMonth) {
            LocalDate.now(ZoneOffset.UTC).dayOfMonth
        } else {
            daysInMonth(monthStr)
        }
    }

    /** Hours between a check-in and check-out ISO instant, or null if either is missing/unparseable
     * or out precedes in (e.g. a still-open shift). Real elapsed time, so a night shift crossing
     * midnight is handled automatically — no separate date-rollover logic needed. */
    fun hoursBetween(inIso: String?, outIso: String?): Double? {
        if (inIso.isNullOrBlank() || outIso.isNullOrBlank()) return null
        return try {
            val seconds = java.time.Duration.between(Instant.parse(inIso), Instant.parse(outIso)).seconds
            if (seconds < 0) null else seconds / 3600.0
        } catch (e: DateTimeParseException) {
            null
        }
    }

    /** Builds an ISO instant from a calendar date + a local HH:mm time — used by the admin's
     * manual attendance correction dialog, where the picker only knows a clock time. */
    fun isoFromLocalTime(dateStr: String, hour: Int, minute: Int, plusDays: Int = 0): String {
        val date = LocalDate.parse(dateStr, ISO_DATE).plusDays(plusDays.toLong())
        val local = LocalDateTime.of(date, LocalTime.of(hour, minute))
        return local.atZone(ZoneId.systemDefault()).toInstant().toString()
    }

    /** The inverse of [isoFromLocalTime] — local hour/minute to pre-fill an edit dialog, or null if unparseable. */
    fun localHourMinute(iso: String?): Pair<Int, Int>? {
        if (iso.isNullOrBlank()) return null
        return try {
            val local = LocalDateTime.ofInstant(Instant.parse(iso), ZoneId.systemDefault())
            local.hour to local.minute
        } catch (e: DateTimeParseException) {
            null
        }
    }

    fun isWithin(dateStr: String, fromDate: String, toDate: String): Boolean {
        val d = LocalDate.parse(dateStr, ISO_DATE)
        val from = LocalDate.parse(fromDate, ISO_DATE)
        val to = LocalDate.parse(toDate, ISO_DATE)
        return !d.isBefore(from) && !d.isAfter(to)
    }

    /**
     * Best-effort "YYYY-MM-DD" from whatever a spreadsheet from another system happens to contain.
     * Returns "" when nothing sensible can be read, so the caller can fall back to a known date
     * rather than inventing one.
     *
     * The ERP prints dates in its own locale and Excel hands some cells back as a serial number,
     * so guessing is unavoidable — but the guessing is bounded here rather than spread across
     * every importer. Day-first is assumed for ambiguous slash dates (01/09/2026 is 1 September),
     * because that is the convention everywhere this is used; an ISO date is passed through
     * untouched, and a value with a four-digit year first is read year-first regardless.
     */
    fun normaliseDate(raw: String?): String {
        val s = raw?.trim().orEmpty()
        if (s.isBlank()) return ""

        // Excel serial date: days since 1899-12-30. Only treat plausibly-dated numbers this way.
        s.toDoubleOrNull()?.let { serial ->
            val days = serial.toLong()
            if (days in 20_000..60_000) {
                return LocalDate.of(1899, 12, 30).plusDays(days).format(ISO_DATE)
            }
            return ""
        }

        // Drop any time part the cell carries alongside the date.
        val datePart = s.split(' ', 'T').first()
        val parts = datePart.split('-', '/', '.').filter { it.isNotBlank() }
        if (parts.size != 3) return tryMonthName(datePart)

        val nums = parts.map { it.toIntOrNull() ?: return tryMonthName(datePart) }
        val (y, m, d) = when {
            parts[0].length == 4 -> Triple(nums[0], nums[1], nums[2])   // 2026-09-01
            parts[2].length == 4 -> Triple(nums[2], nums[1], nums[0])   // 01/09/2026, day first
            else -> Triple(2000 + nums[2], nums[1], nums[0])            // 01/09/26
        }
        return try {
            LocalDate.of(y, m, d).format(ISO_DATE)
        } catch (e: java.time.DateTimeException) {
            ""
        }
    }

    private val MONTHS = listOf(
        "jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec",
    )

    /** Handles the "01-Sep-2026" shape the ERP uses on some reports. */
    private fun tryMonthName(s: String): String {
        val parts = s.split('-', '/', ' ', '.').filter { it.isNotBlank() }
        if (parts.size != 3) return ""
        val monthIdx = parts.indexOfFirst { p -> MONTHS.any { p.lowercase().startsWith(it) } }
        if (monthIdx == -1) return ""
        val month = MONTHS.indexOfFirst { parts[monthIdx].lowercase().startsWith(it) } + 1
        val others = parts.filterIndexed { i, _ -> i != monthIdx }.mapNotNull { it.toIntOrNull() }
        if (others.size != 2) return ""
        val year = others.firstOrNull { it > 31 }?.let { if (it < 100) 2000 + it else it } ?: return ""
        val day = others.first { it <= 31 }
        return try {
            LocalDate.of(year, month, day).format(ISO_DATE)
        } catch (e: java.time.DateTimeException) {
            ""
        }
    }
}

File: web/js/dateUtils.js
// Mirrors app/src/main/java/com/ktc/sitepulse/domain/DateUtils.kt exactly, which itself mirrors
// the original web app's date conventions:
//  - "today" for attendance doc IDs is the UTC calendar date (new Date().toISOString().slice(0,10))
//    — NOT the device's local date.
//  - Night-shift classification uses the LOCAL device hour.
// Keep both conventions exactly as-is so data stays consistent with existing Firestore records.

export function todayStrUtc() {
  return new Date().toISOString().slice(0, 10);
}

export function nowIso() {
  return new Date().toISOString();
}

export function yesterdayOf(dateStr) {
  const d = new Date(`${dateStr}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() - 1);
  return d.toISOString().slice(0, 10);
}

export function localHour() {
  return new Date().getHours();
}

// nightStart/dayStart let a specific site override the company-wide 18:00/05:00 default.
export function shiftFor(hour = localHour(), nightStart = 18, dayStart = 5) {
  return hour >= nightStart || hour < dayStart ? "Night" : "Day";
}

export function formatTimeHm(iso) {
  if (!iso) return "";
  const d = new Date(iso);
  if (isNaN(d)) return "";
  return d.toLocaleTimeString(undefined, { hour: "numeric", minute: "2-digit" });
}

export function formatDateTime(iso) {
  if (!iso) return "";
  const d = new Date(iso);
  if (isNaN(d)) return "";
  return d.toLocaleString(undefined, {
    day: "numeric", month: "short", year: "numeric", hour: "numeric", minute: "2-digit",
  });
}

export function monthOf(dateStr) {
  return dateStr.slice(0, 7);
}

export function monthStart(monthStr) {
  return `${monthStr}-01`;
}

export function monthEndExclusive(monthStr) {
  const [y, m] = monthStr.split("-").map(Number);
  const next = m === 12 ? { y: y + 1, m: 1 } : { y, m: m + 1 };
  return `${next.y}-${String(next.m).padStart(2, "0")}-01`;
}

export function daysInMonth(monthStr) {
  const [y, m] = monthStr.split("-").map(Number);
  return new Date(Date.UTC(y, m, 0)).getUTCDate();
}

export function elapsedDaysFor(monthStr) {
  const currentMonth = monthOf(todayStrUtc());
  if (monthStr === currentMonth) return new Date().getUTCDate();
  return daysInMonth(monthStr);
}

// Hours between a check-in and check-out ISO instant, or null if either is missing/unparseable
// or out precedes in. Real elapsed time, so a night shift crossing midnight needs no extra logic.
export function hoursBetween(inIso, outIso) {
  if (!inIso || !outIso) return null;
  const inD = new Date(inIso), outD = new Date(outIso);
  if (isNaN(inD) || isNaN(outD)) return null;
  const seconds = (outD.getTime() - inD.getTime()) / 1000;
  return seconds < 0 ? null : seconds / 3600;
}

// Builds an ISO instant from a calendar date + a local HH:mm time — used by the admin's manual
// attendance correction dialog, where the picker only knows a clock time.
export function isoFromLocalTime(dateStr, hour, minute, plusDays = 0) {
  const [y, m, d] = dateStr.split("-").map(Number);
  const local = new Date(y, m - 1, d + plusDays, hour, minute, 0, 0);
  return local.toISOString();
}

// The inverse of isoFromLocalTime — local {hour, minute} to pre-fill an edit dialog, or null.
export function localHourMinute(iso) {
  if (!iso) return null;
  const d = new Date(iso);
  if (isNaN(d)) return null;
  return { hour: d.getHours(), minute: d.getMinutes() };
}

export function isWithin(dateStr, fromDate, toDate) {
  return dateStr >= fromDate && dateStr <= toDate;
}
